### PR TITLE
8267904: C2 crash when compile negative Arrays.copyOf length after loop

### PR DESCRIPTION
**Pull request by huishi-hs**
C2 crash when Arrays.copyOf has a negative length after a loop. This happens in release and debug build. Test and hs_err are in JBS.

Crash reason is:
- CastIINode is created in GraphKit::new_array (in AllocateArrayNode::make_ideal_length), Cast array lenght  to range [0, maxint-2]. This is safe when allocation is success and CastIINode 's input control is InitializeNode's proj control.
- In LibraryCallKit::inline_arraycopy, InitializeNode's proj control's use nodes' control is replaced with AllocateArrayNode's input control (in LibraryCallKit::arraycopy_move_allocation_here). This is necessary to move allocation after array copy checks. But this also includes CastIINode.
```
   C->gvn_replace_by(init->proj_out(TypeFunc::Control), alloc->in(0));
```
- CastIINode's control is also adjust to AllocateArrayNode's input control, which is illegal state in laster IGVN phase, casting a negative to [0, maxint-2].
- This cause control and nodes after loop become top and removed. The previous loop has no fall-through edge and crash.

Fix is:
- In LibraryCallKit::arraycopy_move_allocation_here
    - Before replacing init->proj_out(TypeFunc::Control) in, find and replace  CastIINode nodes with original array length.
    - After move allocation node, create CastIINode again if necessary.

Before fix:  node 250 is CastII which should be after InitializeNode.
image

After fix: all arry copy check is performed on original array length node 203
image

New test test/hotspot/jtreg/compiler/c2/TestNegativeArrayCopyAfterLoop.java is added and pass.
Tests performs on Linux X64 and no regression
- Tier1/2/3/hotspot_all_no_apps on release and fastdebug build.
- Tier1/2/3 with option "-XX:-TieredCompilation -Xbatch" on fastdebug build

---------
### Progress
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue
- [x] Change must be properly reviewed

### Issue
 * [JDK-8267904](https://bugs.openjdk.java.net/browse/JDK-8267904): C2 crash when compile negative Arrays.copyOf length after loop

### Reviewers
 * [Roland Westrelin](https://openjdk.java.net/census#roland) (@rwestrel - **Reviewer**) ⚠️ Review applies to ef4465064e555c4adba186668731c2352c279c55
 * [Vladimir Kozlov](https://openjdk.java.net/census#kvn) (@vnkozlov - **Reviewer**)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.java.net/jdk pull/4238/head:pull/4238` \
`$ git checkout pull/4238`

Update a local copy of the PR: \
`$ git checkout pull/4238` \
`$ git pull https://git.openjdk.java.net/jdk pull/4238/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 4238`

View PR using the GUI difftool: \
`$ git pr show -t 4238`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.java.net/jdk/pull/4238.diff">https://git.openjdk.java.net/jdk/pull/4238.diff</a>

</details>

### CODE REVIEW

File: src/hotspot/share/opto/library_call.cpp
Comment (rwestrel) [inline src/hotspot/share/opto/library_call.cpp:4474]: Is this really necessary? Have you seen cases with multiple identical CastII nodes? Or is it to be extra cautious?

**Review by rwestrel — COMMENTED** [review]
Overall fix looks ok to me.

File: src/hotspot/share/opto/library_call.cpp
Comment (huishi-hs) [reply src/hotspot/share/opto/library_call.cpp:4474]: Thanks for your comment!

Multiple identical CastII node do happens and check is added to ensure they are identical CastIINodes. Cases are:
1. CastIINode created in GraphKit::new_array 
2. CastIINode might created before inline_arraycopy if GraphKit::load_array_length is invoked. If new array's length is loaded multiple times before System.arraycopy, there could be multiple CastIINode.

Comment (huishi-hs) [reply src/hotspot/share/opto/library_call.cpp:4474]: simple example
```
        byte[] result = new byte[len];   // first CastII in GraphKit::new_array
        System.arraycopy(path, begin, result, 0, result.length);  // second CastII  when processing "result.length"
```

Comment (rwestrel) [reply src/hotspot/share/opto/library_call.cpp:4474]: Could you use hash_find() instead? hash_find(prev) == hash_find(cur) maybe?

File: src/hotspot/share/opto/library_call.cpp
Comment (rwestrel) [inline src/hotspot/share/opto/library_call.cpp:4456]: I would suggest rephrasing the comment to:
The CastIINode created in GraphKit::new_array (in AllocateArrayNode::make_ideal_length) must stay below the allocation (i.e. is only valid if the allocation succeeds): 1) replace CastIINode with AllocateArrayNode's length here 2) Create CastIINode again once allocation has moved (see below) at the end of this method

Comment (huishi-hs) [reply src/hotspot/share/opto/library_call.cpp:4451]: Thanks @rwestrel !  Debug build check and comment are updated.

For "Could you use hash_find() instead? hash_find(prev) == hash_find(cur) maybe?"

Previous multiple CastIINodes are different as: 
- In GraphKit::new_array, CastIINode is only recorded for igvn but not transformed.
- In GraphKit::load_array_length, CastIINode is applied "_gvn.transform", which set TypeInt's _widen field as 3.
``` 
267  CastII  ===  263  226  [[]]  #int:0..max-3 !jvms: ZipPath::getFileName @ bci:65 (line 102)
272  CastII  ===  263  226  [[ 210 ]]  #int:0..max-3:www !jvms: ZipPath::getFileName @ bci:76 (line 103)
```

Updated commit applies "_gvn.transform" on CastIINode created after array allocation too.

Comment (huishi-hs) [reply src/hotspot/share/opto/library_call.cpp:4451]: Release/fastdebug tier1/2/3 tested.
Fastdebug Tier1/2/3 with option "-XX:-TieredCompilation -Xbatch" also tested.

File: src/hotspot/share/opto/library_call.cpp
Comment (huishi-hs) [reply src/hotspot/share/opto/library_call.cpp:4465]: Done.

File: src/hotspot/share/opto/graphKit.cpp
Comment (rwestrel) [inline src/hotspot/share/opto/graphKit.cpp:3987]: _gvn.set_type_bottom(ccast);
is not needed then. 
record_for_igvn(ccast);
is likely not needed either. No sure the reason there was no call to transform() in the first place.

Comment (huishi-hs) [reply src/hotspot/share/opto/graphKit.cpp:3987]: Thanks Roland! These codes are stble for a long time.

"_gvn.set_type_bottom(ccast);" is not needed as type will be set in "_gvn.transform".
But not sure if it is safe to remove "record_for_igvn".  Could you share more thoughts on it?  

```
eee15b163 hotspot/src/share/vm/opto/graphKit.cpp (Chuck Rasbold           2008-09-17 08:29:17 -0700 3984)       _gvn.set_type_bottom(ccast);
eee15b163 hotspot/src/share/vm/opto/graphKit.cpp (Chuck Rasbold           2008-09-17 08:29:17 -0700 3986)       record_for_igvn(ccast);
```

Comment (huishi-hs) [reply src/hotspot/share/opto/graphKit.cpp:3987]: FYI.  Tier1/2/3 release/fastdebug build pass when removing "_gvn.set_type_bottom(ccast);" and "record_for_igvn(ccast);"

Comment (huishi-hs) [reply src/hotspot/share/opto/graphKit.cpp:3987]: @rwestrel 
for "record_for_igvn" is likely not needed either.  Please help correct my understanding.

In GraphKit::new_array, CastIINode's _range_check_dependency is false and _gvn is not igvn. 
When "_gvn.transform(ccast)" calls CastIINode::Ideal, it either already in for_igvn (goes to "phase->record_for_igvn(this)") or in "post_loop_opts_igvn". 

No need call record_for_igvn again after " _gvn.transform(ccast)".

Comment (rwestrel) [reply src/hotspot/share/opto/graphKit.cpp:3987]: The usual pattern at parse time is to call gvn.transform() and that's it. Later on, during igvn, if an input of a node changes, it's enqueued to be processed for igvn so at parse time, usually, there's no need to call record_for_igvn().
In this case, I think there's a call to record_for_igvn() because there's no call to gvn.transform(). The node has to be transformed at some point. If it's not done during gvn, it has to be during igvn and the cal to record_for_igvn() guarantees that.

Comment (huishi-hs) [reply src/hotspot/share/opto/graphKit.cpp:3985]: Thanks! New commit is added to remove not needed ops.

File: src/hotspot/share/opto/library_call.cpp
Comment (rwestrel) [inline src/hotspot/share/opto/library_call.cpp:4468]: I don't think a call to record_for_igvn() is needed here.

**Review by rwestrel — APPROVED** [review]
Otherwise looks good to me.

File: src/hotspot/share/opto/library_call.cpp
Comment (huishi-hs) [reply src/hotspot/share/opto/library_call.cpp:4468]: yes, record_for_igvn is called in gvn_replace_by.

**Comment by huishi-hs** [comment]
May I have a second review for this PR?

**Review by vnkozlov — APPROVED** [review]

**Comment by huishi-hs** [comment]
Thanks for helping review @vnkozlov @rwestrel !

**Comment by dean-long** [comment]
This PR seems to have caused a failure in one of Oracle's internal tests.  It may need to be backed out.

**Pull request closed without merging**
